### 날씨예측

`-` 아래와 같이 세상의 법칙이 있다고 하자. 
- 맑/맑: 0.4 
- 맑/비: 0.3 
- 비/맑: 0.2
- 비/비: 0.1 

`-` 모든 $t$에 대하여 확률변수 $X_t$를 아래와 같이 정의하자. 
- 맑: $X_t=0$ 
- 비: $X_t=1$

`-` 오늘 (2021년5월8일) 비가 왔다고 치자. 10000일 뒤에도 비가 올 확률은 얼마일까? 

`-` 풀이1

하나의 평행세계에서 관측한 결과

In [16]:
x<-c()
x[1]<-1
for(t in 1:9999){
    if(x[t]==0) x[t+1]<-(runif(1)<3/7)
    else x[t+1]<-(runif(1)<1/3)
}

In [9]:
x[10000]

[1] 1

1000개정도의 평행세계에서 결과를 관측하고 평균을 내보자. 

In [18]:
doctor<-function(){
    x<-c()
    x[1]<-1
    for(t in 1:9999){
        if(x[t]==0) x[t+1]<-(runif(1)<3/7)
        else x[t+1]<-(runif(1)<1/3)
    }
    x
}

In [20]:
X<-rep(0,1000*10000)
dim(X)<-c(10000,1000)

In [24]:
for(omega in 1:1000){
    X[,omega]<-doctor()
}

In [28]:
apply(X,1,mean)

[1] 1.000 0.321 0.407 0.372 0.428 0.386 0.397 0.392 0.406 0.376 0.402 0.375
   [13] 0.389 0.400 0.377 0.403 0.386 0.376 0.383 0.396 0.366 0.392 0.405 0.364
   [25] 0.409 0.383 0.390 0.375 0.400 0.400 0.359 0.398 0.413 0.386 0.390 0.422
   [37] 0.379 0.395 0.381 0.421 0.411 0.395 0.399 0.350 0.405 0.395 0.407 0.394
   [49] 0.370 0.402 0.377 0.400 0.397 0.392 0.383 0.385 0.383 0.388 0.405 0.409
   [61] 0.363 0.393 0.403 0.376 0.384 0.396 0.396 0.375 0.372 0.403 0.382 0.367
   [73] 0.411 0.420 0.386 0.392 0.397 0.405 0.390 0.355 0.412 0.409 0.390 0.385
   [85] 0.387 0.396 0.361 0.370 0.372 0.386 0.395 0.389 0.388 0.369 0.412 0.400
   [97] 0.372 0.385 0.377 0.405 0.378 0.422 0.400 0.400 0.409 0.385 0.400 0.384
  [109] 0.383 0.411 0.394 0.391 0.401 0.388 0.386 0.364 0.396 0.393 0.403 0.401
  [121] 0.398 0.402 0.397 0.402 0.366 0.403 0.399 0.389 0.369 0.386 0.432 0.388
  [133] 0.387 0.382 0.396 0.386 0.383 0.413 0.367 0.371 0.383 0.381 0.403 0.359
  [145] 0.392 0.388 0.386 0.356 0.402 0.379 0.420 0.406 0.383 0.404 0.386 0.363
  [157] 0.421 0.395 0.389 0.411 0.389 0.438 0.404 0.433 0.399 0.400 0.361 0.391
  [169] 0.379 0.383 0.388 0.413 0.357 0.412 0.387 0.399 0.397 0.405 0.376 0.399
  [181] 0.399 0.411 0.377 0.380 0.379 0.388 0.407 0.365 0.381 0.389 0.382 0.407
  [193] 0.396 0.377 0.417 0.383 0.384 0.419 0.400 0.406 0.373 0.423 0.389 0.392
  [205] 0.381 0.367 0.399 0.381 0.415 0.368 0.397 0.375 0.394 0.391 0.396 0.372
  [217] 0.400 0.394 0.414 0.374 0.405 0.407 0.397 0.376 0.417 0.378 0.391 0.364
  [229] 0.378 0.387 0.390 0.398 0.389 0.406 0.418 0.392 0.377 0.397 0.413 0.378
  [241] 0.402 0.380 0.419 0.389 0.383 0.381 0.396 0.402 0.369 0.350 0.398 0.390
  [253] 0.394 0.391 0.417 0.382 0.386 0.384 0.393 0.387 0.384 0.383 0.369 0.418
  [265] 0.390 0.402 0.388 0.383 0.396 0.411 0.384 0.398 0.391 0.409 0.399 0.389
  [277] 0.397 0.380 0.373 0.402 0.379 0.417 0.378 0.386 0.389 0.399 0.395 0.381
  [289] 0.383 0.359 0.394 0.378 0.389 0.384 0.373 0.401 0.380 0.397 0.396 0.399
  [301] 0.435 0.378 0.373 0.379 0.408 0.399 0.406 0.377 0.388 0.399 0.364 0.390
  [313] 0.372 0.379 0.397 0.372 0.399 0.394 0.419 0.372 0.374 0.413 0.398 0.380
  [325] 0.405 0.397 0.397 0.376 0.377 0.426 0.363 0.405 0.387 0.384 0.381 0.403
  [337] 0.392 0.400 0.371 0.400 0.414 0.375 0.404 0.362 0.389 0.363 0.397 0.389
  [349] 0.420 0.371 0.375 0.386 0.403 0.392 0.367 0.390 0.397 0.375 0.417 0.391
  [361] 0.384 0.416 0.365 0.385 0.399 0.377 0.423 0.388 0.393 0.377 0.373 0.363
  [373] 0.392 0.412 0.388 0.398 0.382 0.409 0.377 0.406 0.373 0.393 0.376 0.398
  [385] 0.386 0.417 0.398 0.409 0.422 0.399 0.368 0.375 0.408 0.377 0.371 0.378
  [397] 0.432 0.350 0.415 0.418 0.408 0.384 0.396 0.403 0.405 0.403 0.385 0.390
  [409] 0.368 0.375 0.397 0.412 0.373 0.393 0.409 0.404 0.392 0.379 0.381 0.399
  [421] 0.391 0.407 0.393 0.399 0.375 0.368 0.412 0.385 0.412 0.411 0.384 0.359
  [433] 0.385 0.402 0.398 0.364 0.394 0.391 0.390 0.400 0.377 0.427 0.367 0.399
  [445] 0.398 0.431 0.382 0.399 0.406 0.407 0.399 0.383 0.387 0.397 0.373 0.417
  [457] 0.393 0.397 0.393 0.365 0.438 0.389 0.378 0.422 0.387 0.389 0.382 0.416
  [469] 0.402 0.371 0.404 0.410 0.396 0.403 0.388 0.386 0.383 0.399 0.387 0.421
  [481] 0.401 0.397 0.385 0.382 0.393 0.376 0.391 0.413 0.370 0.406 0.382 0.381
  [493] 0.402 0.371 0.392 0.391 0.388 0.391 0.378 0.397 0.396 0.405 0.419 0.378
  [505] 0.380 0.405 0.420 0.389 0.419 0.408 0.377 0.392 0.369 0.411 0.378 0.390
  [517] 0.395 0.409 0.395 0.377 0.388 0.391 0.388 0.378 0.397 0.388 0.396 0.396
  [529] 0.400 0.415 0.379 0.385 0.396 0.333 0.406 0.391 0.373 0.407 0.393 0.383
  [541] 0.405 0.382 0.413 0.393 0.409 0.410 0.390 0.404 0.431 0.402 0.408 0.415
  [553] 0.433 0.391 0.372 0.389 0.404 0.372 0.397 0.389 0.353 0.393 0.399 0.372
  [565] 0.374 0.360 0.396 0.389 0.392 0.382 0.400 0.379 0.407 0.405 0.387 0.381
  [577] 0.386 0.395 0.411 0.389 0.390 0.416 0.414 0.380 0.358 0.391 0.382 0.433
  [589] 0.380 0.413 0.372 0.412 0.379 0.350 0.389 0.424 0.389 0.402 0.398 0.399
  [6

대충 0.393 정도의 확률로 10000일 뒤에 비가 온다고 볼 수 있다.

`-` 풀이2

수식으로 나타내면 아래와 같다. 

(1) $X_1=1$

(2) $X_2=\begin{cases}0 & wp 2/3 \\ 1 & wp 1/3 \end{cases}$

결국 $X_2$는 아래와 같은 확률질량함수를 가진다. 

|$x_2$|$0$|$1$|
|:-:|:-:|:-:|
|$p(x_2)$|$2/3$|$1/3$|

(3) $X_3$은 어떠할까? 

$P(X_3=1)=P(X_3=1|X_2=0)P(X_2=0)+P(X_3=1|X_2=1)P(X_2=1)$

$=\frac{3}{7}\frac{2}{3}+\frac{1}{3}\frac{1}{3}\approx 0.397$

$P(X_3=0)=P(X_3=0|X_2=0)P(X_2=0)+P(X_3=0|X_2=1)P(X_2=1)$

$=\frac{4}{7}\frac{2}{3}+\frac{2}{3}\frac{1}{3} \approx 0.603$

|$x_3$|$0$|$1$|
|:-:|:-:|:-:|
|$p(x_3)$|$0.603$|$0.397$|

이런식으로 쭉 구하면 된다.. 

그런데 하다보니까 규칙성있게 식을 표현할 수 있음을 알게 되었다 

- $X_1$의 확률분포 
|$x_1$|$0$|$1$|
|:-:|:-:|:-:|
|$p(x_1)$|$0$|$1$|

- $X_2$의 확률분포

$\begin{cases}
P(X_2=0)=P(X_2=0|X_1=0)P(X_1=0)+P(X_2=0|X_1=1)P(X_1=1)\\ 
P(X_2=1)=P(X_2=1|X_1=0)P(X_1=0)+P(X_2=1|X_1=1)P(X_1=1)
\end{cases}
$

정리하면

$\begin{cases}
P(X_2=0)=\frac{4}{7}P(X_1=0)+\frac{2}{3}P(X_1=1)\\ 
P(X_2=1)=\frac{3}{7}P(X_1=0)+\frac{1}{3}P(X_1=1)
\end{cases}
$

다시정리하면 

$\begin{bmatrix}
P(X_2=0) \\ 
P(X_2=1)
\end{bmatrix}=
\begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3} \end{bmatrix}\begin{bmatrix}
P(X_1=0) \\ 
P(X_1=1)
\end{bmatrix}
$

- $X_3$의 확률분포

$\begin{cases}
P(X_3=0)=P(X_3=0|X_2=0)P(X_2=0)+P(X_3=0|X_2=1)P(X_2=1)\\
P(X_3=1)=P(X_3=1|X_2=0)P(X_2=0)+P(X_3=1|X_2=1)P(X_2=1)
\end{cases}$

$\begin{cases}
P(X_3=0)=\frac{4}{7}P(X_2=0)+\frac{2}{3}P(X_2=1)\\
P(X_3=1)=\frac{3}{7}P(X_2=0)+\frac{1}{3}P(X_2=1)
\end{cases}$

$\begin{bmatrix}
P(X_3=0) \\ 
P(X_3=1)
\end{bmatrix}=
\begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3} \end{bmatrix}\begin{bmatrix}
P(X_2=0) \\ 
P(X_2=1)
\end{bmatrix}
$

- ${\bf P}=\begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3}\end{bmatrix}$ 라고하고 , $\mu=\begin{bmatrix} 0 \\ 1 \end{bmatrix} = \begin{bmatrix} P(X_1=0) \\ P(X_1=1)\end{bmatrix}$ 라고 하면 

$\begin{bmatrix}
P(X_2=0) \\ 
P(X_2=1)
\end{bmatrix}=
\begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3} \end{bmatrix}\begin{bmatrix}
P(X_1=0) \\ 
P(X_1=1)
\end{bmatrix}={\bf P}\mu
$

$\begin{bmatrix}
P(X_3=0) \\ 
P(X_3=1)
\end{bmatrix}=
\begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3} \end{bmatrix}\begin{bmatrix}
P(X_2=0) \\ 
P(X_2=1)
\end{bmatrix}= \begin{bmatrix} \frac{4}{7} & \frac{2}{3} \\ \frac{3}{7} & \frac{1}{3} \end{bmatrix} {\bf P}\mu={\bf P}^2\mu
$

이런식으로 표현하다 보면

$\begin{bmatrix}
P(X_{10000}=0) \\ 
P(X_{10000}=1)
\end{bmatrix}={\bf P}^{9999}\mu
$

In [35]:
P<-cbind(c(4/7,3/7),c(2/3,1/3))

In [36]:
P

[,1]      [,2]     
[1,] 0.5714286 0.6666667
[2,] 0.4285714 0.3333333

In [41]:
P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P%*%P

[,1]      [,2]     
[1,] 0.6086957 0.6086957
[2,] 0.3913043 0.3913043

적당히 큰 숫자 이후에는 ${\bf P}$의 거듭제곱이 항상 아래와 같은 매트릭스로 나온다.  

$\begin{bmatrix}
0.6086957 & 0.6086957 \\ 
0.3913043 & 0.3913043`
\end{bmatrix}$

${\bf P}^{9999}$도 대략 아래와 같은 형태라 추론할 수 있다. 

${\bf P}^{9999}=\begin{bmatrix}
0.6086957 & 0.6086957 \\ 
0.3913043 & 0.3913043
\end{bmatrix}$

${\bf P}^{9999}\mu=\begin{bmatrix}
0.6086957 & 0.6086957 \\ 
0.3913043 & 0.3913043
\end{bmatrix}\begin{bmatrix} 0 \\ 1 \end{bmatrix}=\begin{bmatrix} 0.6086957 \\ 0.3913043 \end{bmatrix}$



따라서 $P(X_{10000}=1)=0.3913043$이다. 

즉 27.347년뒤에 비가 올 확률을 39.13043% 이다. 

[$\ast$] 참고로 $\mu=\begin{bmatrix} 0 \\ 1 \end{bmatrix}$가 아닌 일반적인 $\begin{bmatrix} p \\ 1-p \end{bmatrix}$에 대하여서도 위의확률은 동일하다. 즉 

$\begin{bmatrix}
0.6086957 & 0.6086957 \\ 
0.3913043 & 0.3913043
\end{bmatrix}\begin{bmatrix} p \\ 1-p \end{bmatrix}=\begin{bmatrix} 0.6086957p +0.6086957(1-p)  \\ 0.3913043p +0.3913043(1-p)  \end{bmatrix}=\begin{bmatrix} 0.6086957  \\ 0.3913043 \end{bmatrix}$



따라서 초기값 (5월8일에 비가 오든 말든)에 상관없이 27.347년 뒤에 비가 올 확률은 39.13043%이다.  

`-` 풀이3

10000일 뒤에 비가 올 확률이 $p$

(추측) 9999일 뒤에 비가 올 확률도 $p$ 일 것 같다. 

- 9999일 후 맑음 : $1-p$
    - 10000일 후 맑음: $(1-p) \times \frac{4}{7}$
    - 10000일 후 비옴: $(1-p) \times \frac{3}{7}$
- 9999일 후 비옴 : $p$
    - 10000일 후 맑음: $p \times \frac{2}{3}$
    - 10000일 후 비옴: $p \times \frac{1}{3}$

- 결국 10000일후에 비가올 확률은 $(1-p)\times \frac{3}{7} + p \times \frac{1}{3}$ 가 된다. 

그런데 $10000$일후에 비가 올 확률이 $p$라고 하였으므로, 추축이 옳다면 아래가 성립해야 한다. 

$p=(1-p)\times \frac{3}{7} + p \times \frac{1}{3}$

정리하면? 

$p=\frac{9-9p}{21}+\frac{7p}{21}=\frac{9-2p}{21}$

따라서 

$21p=9-2p$

$p=\frac{9}{23}$

In [51]:
9/23

[1] 0.3913043

따라서 $10000$일 뒤에 비가 올 확률은 $p=0.391304347826087$이다. 

`-` 풀이4

In [54]:
mean(doctor())

[1] 0.3892

In [57]:
doctor<-function(){
    x<-c()
    x[1]<-1
    for(t in 1:9999999){
        if(x[t]==0) x[t+1]<-(runif(1)<3/7)
        else x[t+1]<-(runif(1)<1/3)
    }
    x
}

In [58]:
mean(doctor())

[1] 0.3911046

---

`–` 풀이1의 결과와 풀이4의 결과가 일치한다는 것은 ensemble mean 과 time average가 일치한다는 것을 의미한다. 

`-` 이는 큰수의 법칙의 확장판이다. 

`-` 큰수의 법칙은 아래와 같다. 

$X_1,X_2,\dots,X_{9999}$이 $i.i.d.$일때 

$\frac{1}{9999}\sum_{i=1}^{9999}X_i\approx E(X_{10000})=E(X_{9999})=E(X_{9998})=\dots=E(X_1)$

우리는 풀이4를 적용하여서 적절한 답을 구했는데 이는 큰수의 법칙이 적용된 것으로 해석해도 된다. 

그런데 우리의 경우는 $i.i.d$ 가 아니다. (일단 확실히 독립은 아니다.)

따라서 큰수의 법칙이 원래는 성립하지 않아야 한다. (그런데 한다?)

`-` 결국에는 $X_1,X_2,\dots$이 서로 독립이 아니어도 아래의 식이 만족하는 특별한 열(array)이 있다. 

$\frac{1}{T}\sum_{t=1}^{T} X_t \approx E(X_T)$

이러한 특별한 열에는 어떠한 것들이 있는지 관심이 있다. 

이 식은 결국 time average(좌변)와 ensemble mean(우변)이 같다는 것을 의미하기 떄문이다. 

참고로 iid 버전에서는 sample mean이 time average 에 대응한다. 

`-` iid가 아닌 열중에서 time average $\to$ ensemble mean 이 되는 열들은 매우 많다.

우리는 그중에서 마코프체인이라는 특별한 열에만 관심을 가진다. 

`-` 마코프체인은 아래의 성질을 만족하는 확률변수열 $\{X_t\}$를 의미한다. 

$P(X_{t+1}|X_t,X_{t-1},X_{t-2},\dots,X_1)=P(X_{t+1}|X_t)$

수식이 어렵게 되어있는데 아래의 느낌을 기억하면 편리하다. 

```
t+1 시점에 비가 올지 안올지 알고 싶다면, 단지 직전 시점(t시점)에 비가 왔는지 안왔는지에 대한 정보만 있으면 된다. 그저께, 그그저께 등에서 비가 왔는지 안왔는지에 대한 정보는 알고 있을 필요가 없다.  
```

`-` 우리의 날씨예제는 마코프체인이다. 